python/ray/serve/async_goal_manager.py

```diff
@@ -8,34 +8,92 @@
+                return goal_id
+
+    As of now, each external api call as well as goal states are also
+    checkpointed in case of serve controller failure.
+    """
+
     def __init__(self):
-        self._pending_goals: Dict[GoalId, asyncio.Event] = dict()
+        self._pending_goals: Dict[GoalId, asyncio.Future] = dict()
 
     def num_pending_goals(self) -> int:
         return len(self._pending_goals)
 
     def create_goal(self, goal_id: Optional[GoalId] = None) -> GoalId:
+        """
+        Create a new asyncio.Future as goal and return its id. If an id
+        is already given, assign a future to it and return same id.
+        """
         if goal_id is None:
             goal_id = uuid4()
-        self._pending_goals[goal_id] = asyncio.Event()
+
+        self._pending_goals[
+            goal_id] = asyncio.get_running_loop().create_future()
+
         return goal_id
 
-    def complete_goal(self, goal_id: GoalId) -> None:
+    def complete_goal(self,
+                      goal_id: GoalId,
+                      exception: Optional[Exception] = None) -> None:
+        """
+        Mark given goal as completed, if an exception object is provided,
+        set exception instead.
+        """
         logger.debug(f"Completing goal {goal_id}")
-        event = self._pending_goals.pop(goal_id, None)
-        if event:
-            event.set()
+        future = self._pending_goals.pop(goal_id, None)
+        if future:
+            if exception:
+                future.set_exception(exception)
+            else:
+                future.set_result(goal_id)
 
     def check_complete(self, goal_id: GoalId) -> bool:
-        return goal_id not in self._pending_goals
+        """
+        Check if given goal is completed.
+        """
+        if goal_id not in self._pending_goals:
+            return True
+        else:
+            fut = self._pending_goals[goal_id]
+            return fut.done()
 
-    async def wait_for_goal(self, goal_id: GoalId) -> None:
+    async def wait_for_goal(self, goal_id: GoalId) -> GoalId:
+        """
+        Wait for given goal_id to complete by external code calling
+        complete_goal(goal_id), either result or exception could be set.
+
+        Args:
+            goal_id (GoalId): Target goal_id to wait on
+        Returns:
+            goal_id (GoalId): If goal_id finished successfully
+        Raises:
+            exception (Exception): If set by caller of complete_goal() with
+                an exception object
+        """
         start = time.time()
         if goal_id not in self._pending_goals:
             logger.debug(f"Goal {goal_id} not found")
             return
 
-        event = self._pending_goals[goal_id]
-        await event.wait()
+        fut = self._pending_goals[goal_id]
+        await fut
         logger.debug(
             f"Waiting for goal {goal_id} took {time.time() - start} seconds")
+
+        return fut.result()
```

Comment:
Will this line raise an exception if we do `set_exception`? It might be easier for us to just return the exception explicitly instead of raising it here and catching it back in the client. Whatever provides the most user-friendly stacktrace.

Reply:
From docstrings of future:
```
    def result(self):
        """Return the result this future represents.

        If the future has been cancelled, raises CancelledError.  If the
        future's result isn't yet available, raises InvalidStateError.  If
        the future is done and has an exception set, this exception is raised.
        """


        def exception(self):
        """Return the exception that was set on this future.

        The exception (or None if no exception was set) is returned only if
        the future is done.  If the future has been cancelled, raises
        CancelledError.  If the future isn't done yet, raises
        InvalidStateError.
        """
```

I can give a try to see if returning exception here by calling exception() on the returned fut makes it easier to read.

Reply:
Double checked, current implementation will not throw exception as it's wrapped in remote function, similar to how we handled constructor failure exception propagation, so we need a ray.get() at api.py since async goal manager already threw the exception back to us from update()

```
    def _wait_for_goal(self,
                       goal_id: Optional[GoalId],
                       timeout: Optional[float] = None) -> bool:
        if goal_id is None:
            return True

        ready, _ = ray.wait(
            [self._controller.wait_for_goal.remote(goal_id)], timeout=timeout)

        return len(ready) == 1
```

Reply:
Ok, I still think this handling of the exception is really weird... this is going to include a `RayTaskError` and a . We can simplify this by not using the `asyncio.Future` at all and instead just setting a result in the goal manager that is returned to the client. If that result is an exception, the client can choose to raise the exception. That'll have a clean stack trace with just the exception we want to send back to the client, and we won't need to use `asyncio.Future` which requires a running loop.